financial-connections/src/main/java/com/stripe/android/financialconnections/FinancialConnectionsSheetFragment.kt

```diff
@@ -0,0 +1,71 @@
+package com.stripe.android.financialconnections
+
+import android.app.Activity
+import android.content.Intent
+import android.net.Uri
+import android.os.Bundle
+import androidx.activity.addCallback
+import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
+import androidx.fragment.app.Fragment
+import com.airbnb.mvrx.MavericksView
+import com.airbnb.mvrx.activityViewModel
+import com.airbnb.mvrx.withState
+import com.stripe.android.financialconnections.FinancialConnectionsSheetViewEffect.FinishWithResult
+import com.stripe.android.financialconnections.FinancialConnectionsSheetViewEffect.OpenAuthFlowWithUrl
+import com.stripe.android.financialconnections.launcher.FinancialConnectionsSheetActivityResult
+import com.stripe.android.financialconnections.presentation.CreateBrowserIntentForUrl
+
+internal class FinancialConnectionsSheetFragment :
+    Fragment(R.layout.fragment_financial_connections_sheet), MavericksView {
+
+    private val startForResult = registerForActivityResult(StartActivityForResult()) {
+        viewModel.onActivityResult()
+    }
+
+    private val viewModel: FinancialConnectionsSheetViewModel by activityViewModel()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        requireActivity().onBackPressedDispatcher.addCallback(this) {
+            finishWithResult(FinancialConnectionsSheetActivityResult.Canceled)
+        }
+    }
+
+    /**
+     * handle state changes here.
+     */
+    override fun invalidate() {
+        withState(viewModel) { state ->
+            if (state.viewEffect != null) {
+                when (state.viewEffect) {
+                    is OpenAuthFlowWithUrl -> state.viewEffect.launch()
```

Comment:
I think you can directly define the launch() function inside `class OpenAuthFlowWithUrl` instead of an extension, the `when/is` check will be able to tell the type and no downcast is needed here

Reply:
you mean by adding a `launch` abstract function to the `ViewEffect` sealed class? 

each viewEffect uses different dependencies hosted by the fragment, so we can't have a unified `launch`. However I don't think we need that much level of function extraction, we can simplify here removing the launch entirely:  https://github.com/stripe/stripe-android/pull/4986/commits/ea7d683a73a81c0668ed624685b2dcdbf68708f8